src/window.rs

```diff
@@ -0,0 +1,138 @@
+use core::marker::PhantomData;
+use core::f64;
+
+use {Sample, FloatSample};
+use signal;
+use signal::{Phase, ConstHz, Step};
+use conv::{FromSample, ToSample};
+use frame::Frame;
+
+// Using types instead of enum allows for static dispatch
+pub trait WindowType<S: Sample, T: Step> {
+    fn at_phase(phase: S) -> S;
+}
+
+pub struct HanningWindow;
+
+impl<S: Sample + ToSample<f64> + FromSample<f64>, T: Step> WindowType<S, T> for HanningWindow {
+    fn at_phase(phase: S) -> S {
+        let v = phase.to_sample::<f64>() * f64::consts::PI * 2.;
+        (0.5f64 * (1f64 - v.cos())).to_sample::<S>()
+    }
+}
+
+pub struct RectangleWindow;
+
+impl<S: Sample + FromSample<f64>, T: Step> WindowType<S, T> for RectangleWindow {
+    #[allow(unused)]
+    fn at_phase(phase: S) -> S {
+        (1.).to_sample::<S>()
+    }
+}
+
+pub struct Window<S: Sample, T: Step, F: Frame<Sample=S>, WT: WindowType<S, T>> {
+    pub phase: Phase<ConstHz>,
+    stype: PhantomData<S>,
+    ttype: PhantomData<T>,
+    ftype: PhantomData<F>,
+    wttype: PhantomData<WT>
+}
+
+impl<S: Sample, F: Frame<Sample=S>, WT: WindowType<S, ConstHz>> Window<S, ConstHz, F, WT> {
+    pub fn new(len: usize) -> Self {
+        let step = signal::rate(len as f64 - 1.).const_hz(1.);
+        Window {
+            phase: signal::phase(step),
+            stype: PhantomData,
+            ttype: PhantomData,
+            ftype: PhantomData,
+            wttype: PhantomData
+        }
+    }
+}
+
+impl<S: Sample + FromSample<f64>, F: Frame<Sample=S>, T: Step, WT: WindowType<S, T>> Iterator for Window<S, T, F, WT> {
+    type Item = F;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let v = WT::at_phase((self.phase.next_phase()).to_sample::<S>());
+        Some(F::from_fn(|_| v))
+    }
+}
+
+pub struct WindowedFrame<'a, S: Sample + FromSample<f64>, T: Step, F: 'a + Frame<Sample=S>, WT: WindowType<S, T>> {
+    pub data: &'a [F],
+    pub window: Window<S, T, F, WT>,
+    idx: usize
+}
+
+impl<'a, S: Sample + FromSample<f64>, T: Step, F: 'a + Frame<Sample=S>, WT: WindowType<S, T>> WindowedFrame<'a, S, T, F, WT> {
```

Comment:
Here, a `WindowedFrame` is the windowed chunk of `Frame`s right? I think using the term `Frame` in the name is a little ambiguous as `Frame` already has a pretty specific meaning within the crate. Perhaps we could rename this to `Chunk`? Is `window::Chunk` any better d'you think?